src/dune/expander.ml

```diff
@@ -371,23 +371,41 @@ let expand_and_record acc ~map_exe ~dep_kind ~expansion_kind
     match resolve_binary ~loc:(Some loc) t ~prog:s with
     | Error fail -> add_fail acc fail
     | Ok path -> Some (path_exp path) )
-  | Macro (Lib, s) -> (
-    let lib_dep, file = parse_lib_file ~loc s in
-    Resolved_forms.add_lib_dep acc lib_dep dep_kind;
+  | Macro (Lib { lib_exec; lib_private }, s) -> (
+    let lib, file = parse_lib_file ~loc s in
+    Resolved_forms.add_lib_dep acc lib dep_kind;
     match
-      Artifacts.Public_libs.file_of_lib t.lib_artifacts ~loc ~lib:lib_dep ~file
-    with
-    | Ok path -> Some (path_exp path)
-    | Error e -> add_fail acc { fail = (fun () -> raise e) } )
-  | Macro (Libexec, s) -> (
-    let lib_dep, file = parse_lib_file ~loc s in
-    Resolved_forms.add_lib_dep acc lib_dep dep_kind;
-    match
-      Artifacts.Public_libs.file_of_lib t.lib_artifacts ~loc ~lib:lib_dep ~file
+      if lib_private then
+        let open Result.O in
+        let* lib = Lib.DB.resolve (Scope.libs t.scope) (loc, lib) in
+        let current_project_name = Scope.name t.scope
+        and referenced_project_name =
+          Lib.info lib |> Lib_info.status |> Lib_info.Status.project_name
+        in
+        if Some current_project_name = referenced_project_name then
```

Comment:
It's probably better to avoid polymorphic equality here, since Project_name is abstract.